_rawWindowProc should be a static method.

feerrenrut opened the pull request:
<!--
Please fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/wiki/Github-pull-request-template-explanation-and-examples
-->

### Link to issue number:
None

### Summary of the issue:
I ran into this while working on something else, it was confusing enough to cost me some time.
The indentation level for `_rawWindowProc` makes it seem like a class / instance method of `CustomWindow`, however it does not take a `cls` or `self` argument, and does not need access to anything.
It is then registered as a window proc callback.

I don't think this causes any errors, though it is confusing.

### Description of how this pull request fixes the issue:
Make the  _rawWindowProc a static method of CustomWindow. It does not need to be a class or instance method (and can't be), but it is tightly coupled to CustomWindow.

Added type hints for `_wClass` and `className`

### Testing performed:
Ran NVDA with some debug logging to ensure that `_rawWindowProc` is still being called

### Known issues with pull request:

### Change log entry:

Section: New features, Changes, Bug fixes

feerrenrut committed:
Wrong indentation level for_rawWindowProc

It was not marked as a static method, and was not accepting a cls argument.
Likely intended to be a freestanding function.

leonardder commented:
This is definitely meant to be a method on the class, IMO. However, it should have the staticmethod deecorator

feerrenrut commented:
> This is definitely meant to be a method on the class, IMO.

Why do you say this?
- In case you missed it, there an instance method `CustomWindow.windowProc` that this method forwards calls to after looking up the instance based on `hwnd`
- It doesn't take a `cls` or `self`, and has no need for them.
- The first param `hwnd` is an `int`. It is defined this way by `@WNDPROC`  which I think resolved to `WNDPROC=WINFUNCTYPE(LRESULT,HWND,c_uint,WPARAM,LPARAM)`.
- There is no way this can be called with a `cls` argument, so it can't be static.

The main intent for this method seems to be to look up the registered instance with 
`inst = CustomWindow._hwndsToInstances[hwnd]` and forward the windowProc call using `res = inst.windowProc(hwnd, msg, wParam, lParam)`
`_rawWindowProc` has no need to be static or an instance function as far as I can tell.

leonardder commented:
> > This is definitely meant to be a method on the class, IMO.
> 
> Why do you say this?

It is probably due to the fact that I've  been working in C# during the last month, and there as you may know, there are no module level functions, all methods are defined on a class.

The reason why I stated this, is that the _rawWindowProc function is an essential part of the CustomWindow class. Making the function global to the module would be confusing to me, as that suggests that the function is a module function and not specific for the class. Therefore, it's just a matter of feeling that this method belongs to the class, but as a static method.

I think I prefer leaving this as is, with the exception of adding the @staticmethod decorator above @WNDPROC.

leonardder commented on source/windowUtils.py:
```diff
@@ -154,13 +154,15 @@ def __new__(cls, *args, **kwargs):
 	def _get__wClass(cls):
 		return WNDCLASSEXW(
 			cbSize=ctypes.sizeof(WNDCLASSEXW),
-			lpfnWndProc=cls._rawWindowProc,
```
I think the `cls.` part of this line suggests that this is intentional.

feerrenrut committed:
Revert indentation level change

While _rawWindowProc does not need any internal access to CustomWindow,
it is strongly coupled to it. Make _rawWindowProc a static method instead,
to convey this coupling.

Add type hints for _wClass and abstract property className

feerrenrut commented:
I've reverted the indentation change and added `@staticmethod`. I misunderstood your previous comment, thinking of `@classmethod` when you said `@staticmethod`. Functionally it makes no difference whether it is a "private" module level function, or a static method. The static method conveys the coupling with the CustomWindow more clearly so I'm convinced. I'll fix up the description on this PR.

leonardder commented:
Thanks.

Actually, this being a classmethod would be nicer since we would be able to access cls directly when looking up the instance, but unfortunately, the decorator way doesn't work.
